add new /spend/user api endpoint

kaiwend opened the pull request:
👋 I was looking for an endpoint to fetch spend and token usage for a user per model for a certain time range. As I don't want to fetch all the spend log records and then filter/aggregate on the client side, as it might lead to significant performance issues for large amount of spend log entries, I added a new api endpoint with this PR.

It can be found under `/spend/user` and takes `user_id` as required parameter, as well as `start_date` and `end_date` as optional parameters to limit the time range to get the info for.

kaiwend committed:
add new /spend/user api endpoint

ishaan-jaff reviewed (changes requested):
@kaiwend is your goal to get the spend per `end_user` from this endpoint ? meaning the `user` passed to chat.completions ?

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -4866,6 +4866,111 @@ async def spend_key_fn():
+    },
+
+)
+async def spend_per_user_fn(
```
Please can you add a test for this in `test_key_generate_prisma.py` ?

ishaan-jaff reviewed (changes requested):
please can you add a test for this

krrishdholakia commented:
@kaiwend if this for spend per user where user is in LiteLLMUserTable 

then we could work on augmenting `/user/info` which already returns all-up user spend 

cc: @matthiaslau @bcvanmeurs i saw y'all liked this too

kaiwend committed:
add test for /spend/user

kaiwend commented:
> @kaiwend is your goal to get the spend per `end_user` from this endpoint ? meaning the `user` passed to chat.completions ?

I am actually not really sure on the difference between `end_user` and `user`. When I check my local database, the spend log entries have both set to the user id: the `user` as well as `end_user` property. Is there a place where I can read up on this?

Also added some tests

krrishdholakia commented:
user = key owner
end-user = the user passed in via `user` param in /chat/completions call

kaiwend commented:
Then in that case it is the end-user.

kaiwend committed:
rename /spend/user to /spend/end_user

krrishdholakia commented:
where in docs would it help to add this information re: end-users + end-user spend tracking? @kaiwend

kaiwend commented:
Ok, I think I got something mixed up here. I am actually trying to get the spend for the key owning `user` and not the `end_user`. What confused me was that both were set for my spend logs locally. I am actually trying to get the info for the user as it is also passed in to the database query with the filter being:
```python
 filter_query = {"user": end_user_id, ...}
```
Sorry for the confusion, I would go ahead and revert my last commit. Also to give you my use case for some context on this:
Basically I want to have several users creating their own virtual keys. They can then use those keys to make requests to different endpoints. For those requests I want to collect insights on usage (spend, input- and output-tokens) for certain time periods (start_date, end_date) per user per model.

In terms of documentation, I would see it somewhere under `/proxy/virtual_keys#advanced---spend-tracking`. What do you think?

ishaan-jaff commented:
Hi @kaiwend can we hop on a call sometime this week to speed along this PR? Here's a link to my cal for your convenience: https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat?month=2024-04

kaiwend committed:
Revert "rename /spend/user to /spend/end_user"

This reverts commit 9a6de32d3ed97fd4d5091a6f3dd7bb561cc9e6dc.

kaiwend committed:
Move token usage call to spend/users

kaiwend committed:
Merge branch 'main' into litellm_add_spend_user_api_endpoint

kaiwend committed:
fix: Remove reverted endpoint route

kaiwend committed:
fix: /spend/users api definition return schema

kaiwend commented:
Removed the previously proposed `/spend/user` endpoint.
Introduced new query params on the `/spend/users` endpoint: `include_token_usage`, `start_date`, `end_date`. When include_token_usage is set to true, the response includes it for the fetched user/users. 

Response now looks like this:
```json
[
  {
    "user_id": "string",
    "max_budget": 0,
    "spend": 0,
    "model_max_budget": {},
    "model_spend": {},
    "user_email": "string",
    "models": [],
    "tpm_limit": 0,
    "rpm_limit": 0,
    "token_usage": [
      {
        "model": "string",
        "spend": 0,
        "prompt_tokens": 0,
        "completion_tokens": 0
      }
    ]
  }
]
```

kaiwend commented:
@ishaan-jaff did you manage to have a look here? How should we progress on this?
